rosetta/transactions/intent.go

```diff
@@ -0,0 +1,175 @@
+// Copyright 2021 Optakt Labs OÜ
+//
+// Licensed under the Apache License, Version 2.0 (the "License"); you may not
+// use this file except in compliance with the License. You may obtain a copy of
+// the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+// WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+// License for the specific language governing permissions and limitations under
+// the License.
+
+package transactions
+
+import (
+	"fmt"
+	"strings"
+
+	"github.com/onflow/cadence"
+	"github.com/onflow/flow-go/model/flow"
+
+	"github.com/optakt/flow-dps/models/convert"
+	"github.com/optakt/flow-dps/models/dps"
+	"github.com/optakt/flow-dps/rosetta/failure"
+	"github.com/optakt/flow-dps/rosetta/object"
+)
+
+// Intent describes the intent of an array of Rosetta operations.
+type Intent struct {
+	From     flow.Address
+	To       flow.Address
+	Amount   cadence.UFix64
+	Payer    flow.Address
+	Proposer flow.Address
+
+	ReferenceBlock            flow.Identifier
+	ProposerKeySequenceNumber uint64
+	GasLimit                  uint64
+}
+
+// CreateTransactionIntent creates a transaction Intent from two operations given as input.
+// Specified operations should be symmetrical, a deposit and a withdrawal from two
+// different accounts. At the moment, the only fields taken into account are the
+// account IDs, amounts and type of operation.
+func (p *Parser) CreateTransactionIntent(operations []object.Operation) (*Intent, error) {
```

Comment:
This should first check the length of the operations slice. Also, `CreateIntent` should be fine.

Reply:
The length of the operations slice is checked in the `Preprocess` handler before calling this.

For me the plus side of checking it there is that, once it comes here, we know we have exactly two operations and two accounts to tag the `sender` and `receiver` for the `InvalidIntent` error.

Reply:
A function should not make assumptions about its input being correct. If you do that, you put the boundary between components in the wrong place. If you put the check into the function, you can use a sentinel error to return a specific error to the API.